Requirements for native integration with AmazonSQS (#5172)
Particular/docs.particular.net: samples/sqs/native-integration/sample.md

- transports/sqs
---

This sample demonstrates how to enable an NServiceBus endpoint to receive messages sent by a native implementation.

First of all, an external system sends a message to an SQS queue using the Amazon SQS SDK.
The main requirement for NServiceBus to be able to consume this message is the presence of the `MessageTypeFullName` message attribute.
Some random attributes are aded to demonstrate how to access those from handlers or behaviors in the pipeline.

snippet: SendingANativeMessage

On the receiving end, an NServiceBus endpoint is listening to the queue and has a handler in place to handle messages of type `SomeNativeMessage`.

First, it will be intercepted in the incoming logical message context as there is a behavior in place:

snippet: BehaviorAccessingNativeMessage

> To register this behavior:

Reviewer: ```suggestion
The code to register the above behavior:
```

snippet: RegisterBehaviorInPipeline

Then the handler is invoked. The handler code can also access the native message and its attributes.

> Note: The message attribute `MessageTypeFullName` will not be available anymore in the MessageAttributes collection. Instead it is now part of the Headers collection.

Reviewer: Isn't this only the case for messages that are forwarded (going through recoverability or audit)? We do copy the values https://github.com/Particular/NServiceBus.AmazonSQS/blob/master/src/NServiceBus.Transport.SQS/InputQueuePump.cs#L175-L181 but there are still there when the message is handled the first time.

Author: ```suggestion
Note: The message attribute `MessageTypeFullName` might not be available anymore in the MessageAttributes collection in recoverability scenarios. Instead, it will be part of the Headers collection.
```

Author: @danielmarbach You're right, I reframed it in a suggestion, please take a look.
